plugins/serialization/src/main/java/org/qbicc/plugin/serialization/InstanceInterceptor.java

```diff
@@ -0,0 +1,49 @@
+package org.qbicc.plugin.serialization;
+
+import static org.qbicc.graph.atomic.AccessModes.SingleUnshared;
+
+import org.qbicc.context.CompilationContext;
+import org.qbicc.interpreter.VmObject;
+import org.qbicc.type.definition.LoadedTypeDefinition;
+import org.qbicc.type.descriptor.ClassTypeDescriptor;
+
+/**
+ * This class enables VmObject instances to be modified
+ * immediately before they are serialized.
+ */
+class InstanceInterceptor {
+    private final CompilationContext ctxt;
+
+    InstanceInterceptor(CompilationContext ctxt) {
+        this.ctxt = ctxt;
+    }
+
+     VmObject process(VmObject original) {
```

Comment:
I originally designed this API to actually support a more general substitution of one object with another (which is why it returns a `VmObject`), but we don't need that yet and it is subtle to get right.   In particular, this code is running in `LOWER`, not in `ADD`, so we have to be very careful to not make additional methods or types reachable via interpretation of code paths we didn't execute while constructing the build time heap.